SI-7009: `@throws` annotation synthesized incorrectly

gkossakowski opened the pull request:
The 990b3c7 made `scala.throws` annotation polymorphic but forgot to
adapt compiler code that synthesizes it, e.g. when parsing class files.

The consequence was that we would get non-deterministically either
`scala.throws` or `scala.throws[T]` as a type for synthesized annotation.
The reason is that `Symbol.addAnnotation` would call `tpe` method which
does not initialization of symbol so type parameters list would not be
determined correctly. Only if info of that symbol was forced for other
reason we would get `scala.throws[T]`. That non-deterministic behavior
was observed in sbt's incremental compiler.

Here's list of changes this commit introduces:
- The `Symbol.addAnnotation` that takes symbol as argument calls
  `typeSymbol` instead of `tpe` for deterministic results
- Introduce `Symbol.addAnnotation` overload that allows us to pass
  an applied type
- Change all places where polymorphic annotations are synthesized
  to pass an applied type

Fixes SI-7009.

Review by @adriaanm

gkossakowski commented:
In my excitement that it was (relatively) easy to fix this bug with help of @adriaanm I totally forgot about tests. The only way I can think of testing this is to use macros to access information about annotations for types coming from Java and print it. Do we want to go that far?

paulp commented:
Much of this patch is no-op. There is no difference between these except characters killed in the making of them:

```
// version 1
appliedType(ThrowsClass, c.tpe)
// version 2
appliedType(definitions.ThrowsClass.typeConstructor, List(c.tpe))
```

More importantly, this moves the responsibility for not screwing this up in the wrong  (outward) direction. You certainly do not want every call site to have to know they need to call "typeConstructor". Let appliedType do it - it already does.

gkossakowski commented:
@paulp: thanks. I'll update my PR. Tests (based on `ReplTest`) are coming too.

gkossakowski committed:
Test case for SI-7009.

The next commit fixes the problem itself and it's easier to see
in diff what's being fixed exactly.

gkossakowski commented:
@paulp: can you have another look?

paulp commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -1042,8 +1042,8 @@ abstract class ClassfileParser {
       for (n <- 0 until nClasses) {
-        val cls = pool.getClassSymbol(in.nextChar.toInt)
-        sym.addAnnotation(definitions.ThrowsClass, Literal(Constant(cls.tpe)))
+        val cls = pool.getClassSymbol(in.nextChar.toInt).typeConstructor
```
You don't mean typeConstructor here, even though it will usually work. An argument to appliedType is ill-kinded unless the corresponding parameter has the same kind arity. In this case that means it will work with monomorphic types and fail otherwise, since you are stripping any type arguments off whatever is T in Throwable[T].

```
// package a { class Bippy[T] extends Throwable }

scala> typeOf[a.Bippy[Int]]
res1: $r.intp.global.Type = a.Bippy[Int]

scala> appliedType(ThrowableClass, res1)
res2: $r.intp.global.Type = Throwable[a.Bippy[Int]]

// here is where you are stripping the [Int] off of Bippy
scala> appliedType(ThrowableClass, res1.typeConstructor)
res3: $r.intp.global.Type = Throwable[a.Bippy]
```

gkossakowski replied:
Good catch. I promise to never, ever commit and push code just _before_ going home. That never works.

gkossakowski replied:
Fixed that mistake.

paulp commented on src/reflect/scala/reflect/internal/Symbols.scala:
```diff
@@ -1584,7 +1584,11 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
     // Convenience for the overwhelmingly common case
     def addAnnotation(sym: Symbol, args: Tree*): this.type =
-      addAnnotation(AnnotationInfo(sym.tpe, args.toList, Nil))
+      addAnnotation(AnnotationInfo(sym.typeConstructor, args.toList, Nil))
```
If you get different behavior from sym.tpe_\* and sym.typeConstructor on a monomorphic type, it's a bug. (You should never call .tpe, but it forwards to tpe_*.) So if this is not a no-op I would like to see the test where the difference is on display.

If sym isn't monormophic, then calling typeConstructor is only exchanging one bug for another. If you really want to call typeConstructor, preface it with

```
(assert sym.typeParams.isEmpty, sym)
```

gkossakowski replied:
@paulp: This is meant to deal with polymorphic types. Why would we exchange one bug for another here?

//cc @adriaanm

paulp replied:
Because it is assigning the annotation a malformed type. Either sym is monomorphic, in which case there is no difference between calling .tpe_\* and .typeConstructor, or it requires type arguments, in which case calling .typeConstructor does not create a proper type. At least calling .tpe_\* gives you something which is not ill-kinded, though it is still an invalid type because it has dummy type arguments.

The long and the short of it is that if you are trying to create an AnnotationInfo based on a polymorphic class symbol, the bug is in your rear view mirror. There's no right thing to do, but tpe_\* is less wrong than typeConstructor.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1577/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2304/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1577/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2304/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1577/<br>![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2304/<br>![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

gkossakowski committed:
Test case for SI-7009.

The next commit fixes the problem itself and it's easier to see
in diff what's being fixed exactly.

gkossakowski committed:
SI-7009: `@throws` annotation synthesized incorrectly

The 990b3c7 made `scala.throws` annotation polymorphic but forgot to
adapt compiler code that synthesizes it, e.g. when parsing class files.

The consequence was that we would get non-deterministically either
`scala.throws` or `scala.throws[T]` as a type for synthesized annotation.
The reason is that `Symbol.addAnnotation` would call `tpe` method which
does not initialization of symbol so type parameters list would not be
determined correctly. Only if info of that symbol was forced for other
reason we would get `scala.throws[T]`. That non-deterministic behavior
was observed in sbt's incremental compiler.

Another problem we have is that Scala allows polymorphic exceptions
so in ClassfileParser we could synthesize `@throws` annotation with
wrong (polymorphic) type applied. In such case the best we can do
is to convert such type to monomorphic one by introducing existantials.

Here's list of changes this commit introduces:
  * The `Symbol.addAnnotation` that takes symbol as argument asserts
    that the type represented by that symbol is monomorpjhic
  * Introduce `Symbol.addAnnotation` overload that allows us to pass
    an applied type
  * Change all places where polymorphic annotations are synthesized
    to pass an applied type
  * Handle polymorphic exception types in
    `ClassfileParser.parseExceptions`

Fixes SI-7009.

Review by @adriaanm and @paulp

gkossakowski commented:
Damn. I got the rebase wrong. I'll fix it in a minute.

gkossakowski commented:
Actually, don't bother looking into that. I found more issues. I didn't think that would turn out to be such a big can of worms.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2325/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1600/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2325/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1600/

paulp commented:
You might also want to look at #1901 which appears to have lots of overlap with this. I am ready for that one to be merged, it has been waiting a while.

paulp commented:
Let's close this pending greater confidence in its correctness.

gkossakowski commented:
Yup. Should be closed. I'll reopen once I figure out remaining bits.
